include/libtorrent/ssl/stream.hpp

```diff
@@ -0,0 +1,217 @@
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_SSL_SSL_STREAM_HPP_INCLUDED
+#define TORRENT_SSL_SSL_STREAM_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+
+#ifdef TORRENT_USE_SSL
+
+#include "libtorrent/ssl/context.hpp"
+
+#include <boost/asio.hpp>
+#include <boost/system/error_code.hpp>
+
+#ifdef TORRENT_USE_OPENSSL
+#include <boost/asio/ssl/stream.hpp>
+#endif
+
+#ifdef TORRENT_USE_GNUTLS
+#include <boost/asio/gnutls/stream.hpp>
+#include <boost/asio/gnutls/stream_base.hpp>
+#endif
+
+#ifndef BOOST_NO_EXCEPTIONS
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include <boost/system/system_error.hpp>
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+#endif
+
+#include <limits>
+#include <string>
+#include <cstddef>
+#include <utility>
+#include <list>
+#include <iterator>
+#include <sstream>
+
+namespace libtorrent {
+namespace ssl {
+
+using boost::system::error_code;
+
+#ifdef TORRENT_USE_OPENSSL
+
+namespace base = boost::asio::ssl;
+
+class TORRENT_EXTRA_EXPORT stream_base
+{
+public:
+	using native_handle_type = SSL*;
+	using handshake_type = base::stream_base::handshake_type;
+	static constexpr auto client = handshake_type::client;
+	static constexpr auto server = handshake_type::server;
+
+	stream_base(context& ctx) : m_context_impl(ctx.m_impl) {}
+	stream_base(stream_base&& other) : m_context_impl(std::move(other.m_context_impl)) {}
+	stream_base(stream_base const &other) = delete;
+	virtual ~stream_base() = default;
+
+	virtual native_handle_type native_handle() = 0;
+
+	context& get_context() const
+	{
+		if(!m_context_impl->parent)
+			throw std::logic_error("access to destroyed ssl context");
+
+		return *m_context_impl->parent;
+	}
+
+	void set_context(context& ctx)
+	{
+		if(m_context_impl != ctx.m_impl)
+		{
+			SSL_set_SSL_CTX(native_handle(), ctx.native_handle());
+			SSL_set_verify(native_handle()
+					, SSL_CTX_get_verify_mode(ctx.native_handle())
+					, SSL_CTX_get_verify_callback(ctx.native_handle()));
+
+		}
+		m_context_impl = ctx.m_impl;
+	}
+
+protected:
+	void init_base();
+
+private:
+	static int ex_index;
+	std::shared_ptr<context::impl> m_context_impl;
+	friend class context;
+};
+
+template<typename NextLayer>
+class stream : public base::stream<NextLayer>, public stream_base
```

Comment:
is there a reason `stream` cannot be a typedef to either `boost::asio::ssl::context` or the gnutls counterpart?

Reply:
The only reason is the implementation of `get_context` in that case. However, I think it's doable by storing the reference to it in ex data of the corresponding `SSL_CTX`.